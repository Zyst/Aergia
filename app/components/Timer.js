// @flow
import React, { Component } from "react";

type Props = {
  active: boolean,
  minutes: number
};

type State = {
  time: number
};

class Timer extends Component {
  props: Props;
  state: State;

  static defaultProps = {
    active: false
  };

  constructor(props: Props) {
    super(props);

    this.state = {
      // Time is in seconds
      time: this.props.minutes * 60
    };
  }

  /**
   * Calls itself continuously as long as we are active, and handles
   * reducing time and making sure our state is "Accurate"
   *
   * @memberof Timer
   */
  handleTimer(): void {
    if (this.props.active) {
      // We call ourselves again
      setTimeout(this.handleTimer(), 1000);

      // And we reduce time
      this.reduceTime();

      // We check if we need to reset our time
    } else if (this.state.time !== this.props.minutes * 60) {
      // We set back the time state to the initial value
      this.setState({ time: this.props.minutes * 60 });
    }
  }

  reduceTime(): void {
    if (this.state.time > 0) {
      this.setState({ time: this.state.time - 1 });
    }
  }

  render() {
    // On render we start to see if our timer needs to be handled
    this.handleTimer();

    return (
      <h1 className="timer">
        {displayTime(this.state.time)}
      </h1>
    );
  }
}

/**
 * Gets time in seconds, returns a formatted string which we'll use to display
 * the timer
 *
 * @export
 * @param {number} time in seconds
 * @returns {string}
 */
export function displayTime(time: number): string {
  const hours = Math.floor(time / 3600);

  const remaining = time - hours * 3600;

  const minutes = Math.floor(remaining / 60);
  const seconds = remaining - minutes * 60;

  if (hours > 0)
    return `${hours}:${zeroPadding(minutes)}:${zeroPadding(seconds)}`;
  return `${zeroPadding(minutes)}:${zeroPadding(seconds)}`;
}

export function zeroPadding(seconds: number): string {
  const secondsString = String(seconds);

  if (secondsString.length > 1) {
    // Basically do nothing
    return secondsString;
  }
  return `0${secondsString}`;
}

export default Timer;
