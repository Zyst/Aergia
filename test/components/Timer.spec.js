// import { spy } from "sinon";
import React from "react";
import { shallow } from "enzyme";
// import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import Timer, { zeroPadding, displayTime } from "../../app/components/Timer";

const setup = () => {
  const component = shallow(<Timer active minutes={25} />);

  return {
    component,
    timer: component.find(".timer")
  };
};

describe("Timer component", () => {
  it("Should be defined", () => {
    expect(Timer).toBeDefined();
  });

  it("Should display a timer", () => {
    const { timer } = setup();

    expect(timer.text()).toBe("25:00");
  });

  it("Should match snapshot", () => {
    const tree = renderer.create(<Timer active minutes={15} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("Should have a timerInterval when active", () => {
    const { component } = setup();

    // From: https://html.spec.whatwg.org/multipage/timers-and-user-prompts.html#timer-initialisation-steps
    // If previous handle was provided, let handle be previous handle; otherwise, let handle be a user-agent-defined integer that is greater than zero that will identify the timeout to be set by this call in the list of active timers.
    expect(component.state().timerInterval).toBeGreaterThan(0);
  });

  it("Should have a timerInterval of -1 when inactive", () => {
    const component = shallow(<Timer minutes={10} />);

    expect(component.state().timerInterval).toBe(-1);
  });

  describe("Zero padding", () => {
    it("Should be defined", () => {
      expect(zeroPadding).toBeDefined();
    });

    it("Should pad zeroes", () => {
      expect(zeroPadding(0)).toBe("00");
      expect(zeroPadding(9)).toBe("09");
    });

    it("Should ignore numbers that don't need to be padded", () => {
      expect(zeroPadding(10)).toBe("10");
      expect(zeroPadding(1992)).toBe("1992");
    });
  });

  describe("Display time", () => {
    it("Should be defined", () => {
      expect(displayTime).toBeDefined();
    });

    it("Should return seconds as seconds with 00 minutes", () => {
      expect(displayTime(0)).toBe("00:00");
      expect(displayTime(59)).toBe("00:59");
    });

    it("Should return minutes properly", () => {
      expect(displayTime(3540)).toBe("59:00");
      expect(displayTime(3599)).toBe("59:59");
    });

    it("Should return hours properly (No Zero pad)", () => {
      expect(displayTime(3666)).toBe("1:01:06");
      expect(displayTime(36000)).toBe("10:00:00");
    });
  });
});
