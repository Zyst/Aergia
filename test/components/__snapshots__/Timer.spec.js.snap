// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Timer component Should match snapshot 1`] = `
<h1
  className="timer"
>
  15:00
</h1>
`;
